packages/@react-stately/color/src/useColorAreaState.ts

```diff
@@ -0,0 +1,217 @@
+  }
+
+  if (isNaN(yChannelStep)) {
+    yChannelStep = color.getChannelRange(channels.yChannel).step;
+  }
+
+  let [isDragging, setDragging] = useState(false);
+  let isDraggingRef = useRef(false).current;
+
+  let xValue = color.getChannelValue(channels.xChannel);
+  let yValue = color.getChannelValue(channels.yChannel);
+  let setXValue = (v: number) => {
+    valueRef.current = color.withChannelValue(channels.xChannel, v);
+    setColor(valueRef.current);
+  };
+  let setYValue = (v: number) => {
+    valueRef.current = color.withChannelValue(channels.yChannel, v);
+    setColor(valueRef.current);
+  };
+  let xChannelPageStep = Math.max(color.getChannelRange(channels.xChannel).pageSize, xChannelStep);
+  let yChannelPageStep = Math.max(color.getChannelRange(channels.yChannel).pageSize, yChannelStep);
+
+  return {
+    channels,
+    xChannelStep,
+    yChannelStep,
+    xChannelPageStep,
+    yChannelPageStep,
+    value: color,
+    setValue(value) {
+      let c = normalizeColor(value);
+      valueRef.current = c;
+      setColor(c);
+    },
+    xValue,
+    setXValue,
+    yValue,
+    setYValue,
+    setColorFromPoint(x: number, y: number) {
+      let {minValue: minValueX, maxValue: maxValueX} = color.getChannelRange(channels.xChannel);
+      let {minValue: minValueY, maxValue: maxValueY} = color.getChannelRange(channels.yChannel);
+      let newXValue = minValueX + clamp(x, 0, 1) * (maxValueX - minValueX);
+      let newYValue = minValueY + (1 - clamp(y, 0, 1)) * (maxValueY - minValueY);
+      let newColor:Color;
+      if (newXValue !== xValue) {
+        // Round new value to multiple of step, clamp value between min and max
+        newXValue = snapValueToStep(newXValue, minValueX, maxValueX, xChannelStep);
+        newColor = color.withChannelValue(channels.xChannel, newXValue);
+      }
+      if (newYValue !== yValue) {
+        // Round new value to multiple of step, clamp value between min and max
+        newYValue = snapValueToStep(newYValue, minValueY, maxValueY, yChannelStep);
+        newColor = (newColor || color).withChannelValue(channels.yChannel, newYValue);
+      }
+      if (newColor) {
+        setColor(newColor);
+      }
+    },
+    getThumbPosition() {
+      let {minValue: minValueX, maxValue: maxValueX} = color.getChannelRange(channels.xChannel);
+      let {minValue: minValueY, maxValue: maxValueY} = color.getChannelRange(channels.yChannel);
+      let x = (xValue - minValueX) / (maxValueX - minValueX);
+      let y = 1 - (yValue - minValueY) / (maxValueY - minValueY);
+      return {x, y};
+    },
+    incrementX(stepSize) {
+      let range = color.getChannelRange(channels.xChannel);
+      setXValue(snapValueToStep(xValue + stepSize, range.minValue, range.maxValue, stepSize));
+    },
+    incrementY(stepSize) {
+      let range = color.getChannelRange(channels.yChannel);
+      setYValue(snapValueToStep(yValue + stepSize, range.minValue, range.maxValue, stepSize));
+    },
```

Comment:
A similar thing happens with default arrow key navigation for the story w/ step size 16

Reply:
ok, so this has to do with how rgb colors work, they are 0-255, and steps such as 16, would go 0, 16, 32, ...240, 256.
however, 256 is outside of the range, and 255 isn't on the step. If I allow it to go to the edge, then the last step is only 15, which is a little weird? it's the difference between 0xF and 0x10, one of them goes evenly into 0xFFFFF and the other doesn't.
Maybe the default step value should be 0xF instead I think if I write them all as hex it'll make more sense

Reply:
Hm, I guess the behavior is fine if the step size is explicitly defined (kinda weird that it effectively prevents moving the thumb past a certain point but I guess that is on the user since they defined the step size)... For the default case where step is 1, I think shift+arrow keys or pgUp/Down should still jump to the edge instead of adhering to the step value and getting stuck.

Perhaps we can discuss and punt on it for this PR. I can def see the argument where the step size should always be respected since native input sliders behave that way

Reply:
I fix this with https://github.com/adobe/react-spectrum/pull/2570/files#diff-3df80adf3ed96502f6c411a5fa83c3e78818de38e2c67de29fbf9582a9fe98aaR229-R234.

Reply:
```ts
    incrementX(stepSize) {
      setXValue(xValue + stepSize > maxValueX ? maxValueX : snapValueToStep(xValue + stepSize, minValueX, maxValueX, stepSize));
    },
    incrementY(stepSize) {
      setYValue(yValue + stepSize > maxValueY ? maxValueY : snapValueToStep(yValue + stepSize, minValueY, maxValueY, stepSize));
    },
```